### PR TITLE
[FLINK-30860][doc] Add document for hybrid shuffle with adaptive batch scheduler

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Add document for hybrid shuffle with adaptive batch scheduler.*

## Brief change log

  - *Add document for hybrid shuffle with adaptive batch scheduler*
  - *Add document for hybrid shuffle index spilling*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/content.zh/docs/ops/batch/batch_shuffle.md [hunk 1]
```diff
@@ -114,12 +114,34 @@ Hybrid shuffle provides two spilling strategies:
 
 To use hybrid shuffle mode, you need to configure the [execution.batch-shuffle-mode]({{< ref "docs/deployment/config" >}}#execution-batch-shuffle-mode) to `ALL_EXCHANGES_HYBRID_FULL` (full spilling strategy) or `ALL_EXCHANGES_HYBRID_SELECTIVE` (selective spilling strategy).
 
+#### Supports AdaptiveBatchScheduler and SpeculativeExecution
+
+Hybrid shuffle currently supports `AdaptiveBatchScheduler` by default. If you want to use `DefaultScheduler`, please configure the [jobmanager.scheduler]({{< ref "docs/deployment/config" >}}#jobmanager-scheduler) to `DefaultScheduler`. See [elastic_scaling]({{< ref "docs/deployment/elastic_scaling" >}}#adaptive-batch-scheduler) for details.
+
+If you want to enable `SpeculativeExecution` in the same time, see [speculative_execution]({{< ref "docs/deployment/speculative_execution" >}}) for details.
```
Comment (xintongsong) [inline docs/content.zh/docs/ops/batch/batch_shuffle.md:121]: This is irrelevant to hybrid shuffle.

File: docs/content.zh/docs/ops/batch/batch_shuffle.md [hunk 2]
```diff
@@ -114,12 +114,34 @@ Hybrid shuffle provides two spilling strategies:
...
+
+Hybrid shuffle divides the partition data consumption constraints between producer and consumer into the following three cases:
+
+- **ALL_PRODUCERS_FINISHED** : hybrid partition data can be consumed only when all producers are finished.
+- **ONLY_FINISHED_PRODUCERS** : hybrid partition data can be consumed when its producer is finished.
+- **UNFINISHED_PRODUCERS** : hybrid partition data can be consumed even if its producer is un-finished.
+
+If `SpeculativeExecution` is enabled, the default constraint is `ONLY_FINISHED_PRODUCERS` to bring some performance optimization compared with blocking shuffle. Otherwise, the default constraint is `UNFINISHED_PRODUCERS` to perform pipelined-like shuffle. These could be configured via [jobmanager.partition.hybrid.partition-data-consume-constraint]({{< ref "docs/deployment/config" >}}#jobmanager-partition-hybrid-partition-data-consume-constraint).
```
Comment (xintongsong) [inline docs/content.zh/docs/ops/batch/batch_shuffle.md:129]: What is the potential impacts when changing this option?

File: docs/content.zh/docs/ops/batch/batch_shuffle.md [hunk 3]
```diff
@@ -114,12 +114,34 @@ Hybrid shuffle provides two spilling strategies:
...
+
+#### Index Spilling
+
+Hybrid shuffle indexes the shuffle data in memory and disk. Generally speaking, all index can be cached in memory to speed up index retrieval. However, for large batch jobs, this part of memory may bring OOM risks.
+Therefore, hybrid shuffle supports spilling index data to disk. The following configuration options can control this behavior:
+
+- **[taskmanager.network.hybrid-shuffle.num-retained-in-memory-regions-max]({{< ref "docs/deployment/config" >}}#taskmanager-network-hybrid-shuffle-num-retained-in-memory-regions-max)** : Controls the max number of hybrid retained regions in memory. Increasing this value will allow more index entries to be cached in memory.
+- **[taskmanager.network.hybrid-shuffle.spill-index-segment-size]({{< ref "docs/deployment/config" >}}#taskmanager-network-hybrid-shuffle-spill-index-segment-size)** : Controls the segment size(in bytes) of hybrid spilled file data index.
```
Comment (xintongsong) [inline docs/content.zh/docs/ops/batch/batch_shuffle.md:137]: I wonder if we want to advertise these options. They are likely removed in future releases.

Comment (reswqa) [reply docs/content.zh/docs/ops/batch/batch_shuffle.md:137]: Agree with you, removed.

File: docs/content.zh/docs/ops/batch/batch_shuffle.md (hunk 2 above)
Comment (reswqa) [reply docs/content.zh/docs/ops/batch/batch_shuffle.md:129]: Added some descriptions of this part.

File: docs/content.zh/docs/ops/batch/batch_shuffle.md (hunk 1 above)
Comment (reswqa) [reply docs/content.zh/docs/ops/batch/batch_shuffle.md:121]: Removed.

**Comment by reswqa** [comment]
Thanks @xintongsong for the review, I have addressed all comments, PTAL~

File: docs/content/docs/ops/batch/batch_shuffle.md [hunk 4]
```diff
@@ -112,14 +112,27 @@ Hybrid shuffle provides two spilling strategies:
...
-To use hybrid shuffle mode, you need to configure the [execution.batch-shuffle-mode]({{< ref "docs/deployment/config" >}}#execution.batch-shuffle-mode) to `ALL_EXCHANGES_HYBRID_FULL` (full spilling strategy) or `ALL_EXCHANGES_HYBRID_SELECTIVE` (selective spilling strategy).
+To use hybrid shuffle mode, you need to configure the [execution.batch-shuffle-mode]({{< ref "docs/deployment/config" >}}#execution-batch-shuffle-mode) to `ALL_EXCHANGES_HYBRID_FULL` (full spilling strategy) or `ALL_EXCHANGES_HYBRID_SELECTIVE` (selective spilling strategy).
+
+#### Data Consumption Constraints
+
+Hybrid shuffle divides the partition data consumption constraints between producer and consumer into the following three cases:
+
+- **ALL_PRODUCERS_FINISHED** : hybrid partition data can be consumed only when all producers are finished.
+- **ONLY_FINISHED_PRODUCERS** : hybrid partition data can be consumed when its producer is finished.
+- **UNFINISHED_PRODUCERS** : hybrid partition data can be consumed even if its producer is un-finished.
```
Comment (xintongsong) [inline docs/content/docs/ops/batch/batch_shuffle.md:123]:
```suggestion
- **ONLY_FINISHED_PRODUCERS** : hybrid partition can only consume data from finished producers.
- **UNFINISHED_PRODUCERS** : hybrid partition can consume data from unfinished producers.
```

File: docs/content/docs/ops/batch/batch_shuffle.md [hunk 5]
```diff
@@ -112,14 +112,27 @@ Hybrid shuffle provides two spilling strategies:
...
 
 ### Limitations
 
 Hybrid shuffle mode is still experimental and has some known limitations, which the Flink community is still working on eliminating.
 
 - **No support for Slot Sharing.** In hybrid shuffle mode, Flink currently forces each task to be executed in a dedicated slot exclusively. If slot sharing is explicitly specified, an error will occur.
-- **No support for Adaptive Batch Scheduler and Speculative Execution.** If adaptive batch scheduler is used in hybrid shuffle mode, an error will occur.
+- **No optimization for dynamic graph.** If auto-parallelism(dynamic graph) is enabled for `AdaptiveBatchScheduler`, hybrid shuffle will always schedule tasks only when all producer are finished like blocking shuffle, this means that the constraint will fall back to `ALL_PRODUCERS_FINISHED` in this case.
```
Comment (xintongsong) [inline docs/content/docs/ops/batch/batch_shuffle.md:135]:
```suggestion
- **No pipelined execution for dynamic graph.** If auto-parallelism (dynamic graph) is enabled, Adaptive Batch Scheduler will wait until upstream tasks finish to decide parallelism of downstream tasks, which means hybrid shuffle effectively fallback to blocking shuffle (`ALL_PRODUCERS_FINISHED` constraint).
```

**Review by xintongsong — APPROVED** [review]
Thanks @reswqa. I have only some minor comments. +1 for merging after addressing these comments.

**Pull request closed without merging**